cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -1818,11 +1818,13 @@ struct ArithmeticFunction : ScalarFunction {
     // Only promote types for binary functions
     if (values->size() == 2) {
       ReplaceNullWithOtherType(values);
-
-      if (auto type = CommonTemporalResolution(values->data(), values->size())) {
+      auto type = CommonTemporalResolution(values->data(), values->size());
+      if (type) {
+        ReplaceTemporalTypes(type, values);
+      } else if (auto numeric_type = CommonNumeric(*values)) {
+        ReplaceTypes(numeric_type, values);
+      } else if (type == TimeUnit::SECOND) {
         ReplaceTemporalTypes(type, values);
```

Comment:
Without adding this we get:
```
'_error_or_value11.status()' failed with NotImplemented: Function 'subtract' has no kernel matching input types (array[timestamp[s]], array[date32[day]])
```
This only happens for (timestamp[s], day32) but not for (timestamp[ms,us,ns], day32) combinations. I'm not quite sure why.

Reply:
@pitrou I'm sure I'm missing something trivial here?

Reply:
Probably because TimeUnit::SECOND is zero :) We should probably change the output type to wrap in`std::optional` or something, or have `bool CommonTemporalResolution(..., TimeUnit::type* out_unit)`